modules/shell/output.go

```diff
@@ -0,0 +1,75 @@
+package shell
+
+import (
+	"strings"
+	"sync"
+)
+
+// output contains the output after runnig a command.
+type output struct {
+	stdout *outputStream
+	stderr *outputStream
+	// merged contains stdout  and stderr merged into one stream.
+	merged *merged
+}
+
+func newOutput() *output {
+	m := new(merged)
+	return &output{
+		merged: m,
+		stdout: &outputStream{
+			merged: m,
+		},
+		stderr: &outputStream{
+			merged: m,
+		},
+	}
+}
+
+func (o *output) Stdout() string {
+	return o.stdout.String()
+}
+
+func (o *output) Stderr() string {
+	return o.stderr.String()
+}
+
+func (o *output) CombinedOutput() string {
+	return o.merged.String()
+}
+
+type outputStream struct {
+	Lines []string
+	*merged
+}
+
+func (st *outputStream) WriteString(s string) (n int, err error) {
+	st.Lines = append(st.Lines, string(s))
+	return st.merged.WriteString(s)
+}
+
+func (st *outputStream) String() string {
+	return strings.Join(st.Lines, "\n")
+}
+
+type merged struct {
+	// ensure that there are no parallel writes
+	sync.Mutex
+	Lines []string
+}
+
+func (m *merged) String() string {
+	m.Lock()
+	defer m.Unlock()
```

Comment:
Why do you need a lock on read?

Reply:
You're right, I don't 🙂 Although this may result in the callee getting "old" data; if the command would still be running. But I don't think that's relevant here.